src/librustc/middle/ty/structural_impls.rs

```diff
@@ -751,6 +658,18 @@ impl<'tcx> TypeFoldable<'tcx> for ty::Predicate<'tcx> {
                 ty::Predicate::ObjectSafe(trait_def_id),
         }
     }
```

Comment:
@nikomatsakis My justification for having `fold_subitems_with` default to `fold_with` was that for types without corresponding methods in `TypeVisitor`, folding the subitems of the type is the same as folding the type.

On second though, I think it would nicest to have `fold_with` default to `fold_subitems_with` so that structural implementations always define `fold_subitem_with` to call `fold_with` on subitems and only define `fold_with` to call a method in `TypeFolder` when there is a corresponding method.

Did you want `fold_subitems_with` to be `unimplemented!` to reflect that users should only be calling `fold_subitems_with` when it is different from `fold_with`? If so, I think it would be better to enforce that at the type level by having a trait like `TypeFolderOverridable` define `fold_subitems_with`, or by using the original `super_*` setup.